python/ray/data/_internal/block_batching/util.py

```diff
@@ -180,6 +180,32 @@ def collate(
         yield CollatedBatch(batch.batch_idx, collated_batch)
 
 
+def finalize_batches(
```

Comment:
can we combine this with the collate function?

Reply:
@amogkam If we want to keep the `finalize_fn()` call in a separate threadpool, then I think we need to keep the logic to apply `finalize_fn()` in a separate function from the collate logic? Otherwise, we cannot separate out collation from the h2d logic in `_async_iter_batches()`: https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/block_batching/iter_batches.py#L147

Let me know if there's a workaround i'm missing here, thanks

Reply:
Hmm you could call make_async_gen twice in the same function right?

```python
def collate_and_finalize(...):
    collated_iter = make_async_gen(base_iter, num_workers=prefetch_batches)
    finalized_iter = make_async_gen(collated_iter, num_workers=1)
    return finalized_iter
```

Reply:
since `make_async_gen` takes `fn=collate` and `fn=finalize_batches`, i think we still need to keep these two functions separate so that they can be run in separate threadpools. as you suggested, i removed `_apply_finalize_fn_in_threadpool` by combining the two `make_async_gen` calls -- thanks!